src/compiler/sys.ts

```diff
@@ -294,6 +294,50 @@ namespace ts {
                 };
             }
 
+            function createWatchedFileSet() {
+                const watchedDirectories: { [path: string]: FileWatcher } = {};
+                const watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    const watchedPaths = Object.keys(watchedDirectories);
+                    // Try to find parent paths that are already watched. If found, don't add directory watchers
+                    const watchedParentPaths = watchedPaths.filter(path => fileName.indexOf(path) === 0);
+                    // If adding new watchers, try to find children paths that are already watched. If found, close them. 
+                    if (watchedParentPaths.length === 0) {
+                        const pathToWatch = ts.getDirectoryPath(fileName);
+                        for (const watchedPath in watchedDirectories) {

```

Comment:
this part is interesting.correct me please if I'm wrong, here you are saying: "As I see nobody is watching folder 'pathToWatch' so I'm going to install watcher for it. However it is possible that somebody is already watching 'pathToWatch/<some-path>'. If yes - I'm going to shutdown this watcher since watcher for 'pathToWatch' should get the same notifications". 

If the answer is 'yes' then I'm curious how it works since by default `fs.watch` is [not recursive](https://nodejs.org/docs/latest/api/fs.html#fs_fs_watch_filename_options_listener) so if initially you had watcher for '/a/b/c/d' and then you install one for '/a/b' then after closing watcher for '/a/b/c/d' you'll receive no notifications about changes in this file